zilliqa/src/db.rs

```diff
@@ -221,6 +235,111 @@ impl Db {
         Ok(())
     }
 
+    pub fn create_checkpoint_path(&self, height: u64) -> Result<Option<Box<Path>>> {
+        let Some(base_path) = &self.path else {
+            // If we don't have on-disk persistency, disable checkpoints too
+            return Ok(None);
+        };
+        let snapshot_path = base_path.join("snapshots").join(height.to_string());
+        fs::create_dir_all(&snapshot_path)?;
+        Ok(Some(snapshot_path.into_boxed_path()))
+    }
+
+    pub fn load_trusted_checkpoint<P: AsRef<Path>>(
+        &self,
+        path: P,
+        hash: &Hash,
+        our_shard_id: u64,
+    ) -> Result<Block> {
+        let input = File::open(path)?;
+        let reader = BufReader::with_capacity(8192 * 1024, input); // 8 MiB read chunks
+        let trie_storage = Arc::new(self.state_trie()?); // TODO: what should we do if the trie DB already contains values?
```

Comment:
Can you clarify what you mean by this?

Reply:
As in, if you're loading a checkpoint but you already have previous state in the DB. For example:
 * You used to run a node, it went out of sync, then you loaded a checkpoint to catch back up (this should be fine). Or...
 * You ran your node on another shard, then switched shards and are loading a checkpoint (maybe bad?)
 * You ran e.g. a testnet and then switched to mainnet and are syncing from a checkpoint without clearing the DB (maybe bad?)
 * You're doing weird stuff, like running a synced node, then maybe deleting/losing/corrupting your sqlite DB (so it no longer records your highest view) and then re-load an earlier checkpoint

Right now this PR just ignores it. Which I _think_ is probably fine assuming there are no hash collisions - I feel like the worst you'll have is a bloated storage with useless values in it. But is even that fine, or do we want to throw a warning saying "your key-value storage seems to have old values that may be useless in it, are you sure?" - since I'd expect the main usecase for checkpoints, I assume, would be to bring up brand new nodes with an empty key-value store.

Reply:
The use case is probably the first bullet point - you have some previous block, your node went down, you skip ahead using a checkpoint. If we disallow previous state, that effectively means you'd have to wipe your node and start fresh if it ever gets out of sync
Which may be fine anyway if people start running pruned nodes as a rule... So I don't have a strong opinion. Mind you the checkpoints only contain the actual state in-use at the checkpoint block height, so old state or otherwise unused bloated state won't propagate through them.